### PR TITLE
Add async interface to store and backend.

### PR DESCRIPTION
**Pull request by lcrh**

### CODE REVIEW

File: src/enact/references.py [hunk 1]
```diff
@@ -242,19 +263,39 @@ def register_type(self,
...
+    """Returns the type, if known."""
+    return self.get_type(type_key)
+
   @abc.abstractmethod
   def commit(self, ref_id: str, packed_resource: PackedResource):
     """Stores a packed resource."""
 
+  async def commit_async(self, ref_id: str, packed_resource: PackedResource):
+    """Stores a packed resource."""
+    return self.commit(ref_id, packed_resource)
```
Comment (stewartmiles) [inline src/enact/references.py:289]: Why return if it doesn't return a value? Or perhaps the return value type annotation is missing?

**Review by stewartmiles — APPROVED** [review]
In general it looks ok, I'm not a huge fan of the duplication.

It would be nice if we could work out an approach that uses a Future / Promise object in all code paths and then the async methods would just wrap the result in an asyncio.Future and be less duplicative. i.e use the Javascript style approach (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises):

e.g ignoring the warnings about `futures.Future` for internal use and the sheer ugliness 🤮 ...

```python
  def commit_concurrent(self, resource: R) -> futures.Future[Ref[R]]:
    """Commits a resource to the store."""
    result_future = futures.Future()
    as_resource = resource_registry.wrap(resource)

    def on_registered_type(registered_future: futures.Future):
      if registered_future.exception():
        result_future.set_exception(registered_future.exception())
        return

      ref, packed_resource = self._ref_type.pack(as_resource)

      def on_backend_commit_complete(
          backend_commit_future: futures.Future):
        if backend_commit_future.exception():
          result_future.set_exception(backend_commit_future.exception())
           return
        result_future.set_result(ref)

      self._backend.commit_concurrent(
        ref.id, packed_resource).add_done_callback(
          on_backend_commit_complete)

    self.register_type_concurrent(as_resource).add_done_callback(
       on_registered_type)
    return result_future

  def commit(self, resource: R) -> Ref[R]:
    """Commits a resource to the store."""
    return self.commit_concurrent(resource).get()

  async def commit_async(self, resource: R) -> Ref[R]:
    """Commits a resource to the store."""
    ref = await asyncio.wrap_future(self.commit_concurrent(resource))
    return ref
```

Wow now I've written it out, it's awful. I guess an alternative would be to just suck it up and start a thread for asyncio and make everything go through that.

File: src/enact/references.py (hunk 1 above)
Comment (lcrh) [reply src/enact/references.py:289]: Fixed.

**Comment by lcrh** [comment]
> In general it looks ok, I'm not a huge fan of the duplication.
> [...]

Yeah I agree. Without having thought about it too much, I'd leaning more towards the other way: The backend should be fully async, the main interface on the store functions is all async and then the sync interface is just a set of wrappers that runs the async functions on an event loop.

**Pull request merged**